### DISCUSSION TITLE
Number of Unread Messages

### DISCUSSION THREAD

**Discussion by justinDeu** [FelixKratz/SketchyBar#503] [Q&A] [answered]
I've been reworking my sketchybar and would really like to have the number of unread text messages from the Messages app appear in the bar. I've looked all over but can't seem to find a way to do this. 

Looking through the AppleScript calls available for Messages, I couldn't see a way to get individual messages and also tell if they were unread to take the count of. I saw some past discussions of looking in `~/Library/Messages/chat.db` and it seems like that is the only possible way unless there are APIs that I haven't found yet. The issue I'm having with the `chat.db` file could be due to the queries I've tried, but it doesn't seem to be updating as I am getting new messages (at least in a timely manner that I can tell) and when filtering the queries to unread messages, there are phantom messages that are read but for some reason are not marked read. Internally they must be using a different query or some other data source to recognize read vs. unread messages because my Messages app does not show these phantom unread messages.

Has anyone come up with a way to do this?

**Comment by FelixKratz** [comment]
Have a look at this comment: https://github.com/FelixKratz/SketchyBar/discussions/317#discussioncomment-4754432
It should give you exactly what you want.

**Reply by justinDeu** [reply]
This looks like it should be exactly what I want, but I can't seem to make it work. When I list all the info, only 3 apps have the StatusLabel property and none are associated with Messages. Looking through entries mentioned Messages, nothing seems indicative of the badge text. For example, right now I have 2 unread texts and in the dock, Messages has the red badge with 2 in it, but using `lsappinfo` I can't seem to identify that information.

**Reply by justinDeu** [reply]
I was able to fix some issues on my computer and get the `sqlite3` method working. It all works as expected when I have unread messages, but there seems to be an issue after I view the messages. The script turns the `drawing` property off for the items to have them disappear. However, this seems to also turn off updating, so the script is not being run to turn back on the `drawing` when more unread messages come in.

In the sketchybar logs I can see the ON messages showing up every 10 seconds until there is the first OFF message then no more appear. 

```bash
# items/messages.sh
sketchybar --add item messages right       \
           --set messages  \
                icon.drawing=off \
	            update_freq=10 \
                background.color=$transparent_color \
                background.padding_left=3 \
                background.padding_right=5 \
                script="~/.config/sketchybar/plugins/messages.sh" \
                click_script="open -a /System/Applications/Messages.app"

sketchybar \
	--add item messages_logo right \
	--set messages_logo icon=􀌤  \
        icon.color=$current_app_color \
        label.drawing=off \
        background.color=$time_highlight \
```

```bash
# plugins/messages.sh 
TEXT=$(sqlite3 ~/Library/Messages/chat.db "SELECT COUNT(guid) FROM message WHERE NOT(is_read) AND NOT(is_from_me) AND text !=''")

if [ $TEXT = 0 ]; then
    echo "$(date): $TEXT -- BAR OFF"
    sketchybar -m \
        --set messages drawing=off \
        --set messages_logo drawing=off
else
    echo "$(date): $TEXT -- BAR ON"
    sketchybar -m \
        --set messages drawing=on label="$TEXT" \
        --set messages_logo drawing=on
fi
```

**Reply by FelixKratz** [reply, marked as answer]
Thats probably because you have set a default for `updates=when_shown`, you need to set it to `on` for this item to continue receiving events even if the item is not shown

**Reply by justinDeu** [reply]
I looked forever trying to figure this out knowing it would be something so simple, thank you so much!
